start-stop-daemon, supervise-daemon: use close_range with musl

ncopa opened the pull request:
Make sure that we use close_range also with musl libc, using syscall directly.

This fixes a severe performance issue when running openrc under docker where the maxfd may be 1G.

thesamesam commented on src/start-stop-daemon/start-stop-daemon.c:
```diff
@@ -191,6 +194,22 @@ static inline int ioprio_set(int which RC_UNUSED,
 }
 #endif
 
+#if !defined(SYS_close_range) && defined(__NR_close_range)
```
I really don't think we should do this. Using `close_range` is fine, but needing hacks for it isn't.

musl should provide a prototype for it?

thesamesam replied:
I think your current approach is alright, even if the situation isn't ideal.

It would be nice if the musl maintainers had replied when you asked about it though :(

Consider my objection withdrawn, thank you for working on this!

ncopa replied:
FWIW, I'm doing a second try of including close_range() in musl: https://www.openwall.com/lists/musl/2023/09/01/1

ncopa commented:
I refactored it a bit to be less hackish. Also changed start-stop-daemon to use CLOEXEC instead of closing the filedescriptors, similar to what supervise-daemon does.

floppym commented on src/shared/misc.c:
```diff
@@ -500,3 +509,29 @@ pid_t get_pid(const char *applet,const char *pidfile)
+#ifdef SYS_close_range
+	return syscall(SYS_close_range, first, last, flags);
+#else
+	return -1;
```
Generally it would be a good idea to set `errno = ENOSYS` before returning.

However, since we don't actually check/log any failures at the call site, I guess it doesn't matter in this specific instance.

ncopa replied:
I agree that `errno = ENOSYS` is a good idea. Let me fix that.

ncopa replied:
Turns out that [syscall(2)](
https://man7.org/linux/man-pages/man2/syscall.2.html) sets `errno`, so we don't need to do it our selves.

> The return value is defined by the system call being invoked.  In general, a 0 return value indicates success.  A -1 return value indicates an error, and an error number is stored in errno.
